src/lib.rs

```diff
@@ -837,6 +837,87 @@ pub enum Format {
     NoSpaces,
 }
 
+/// The representation of a color stop for a `ColorScale`.
+/// The position defines where the color is placed from left (0.0) to right (1.0).
+#[derive(Debug, Clone)]
+struct ColorStop {
+    color: Color,
+    position: Fraction,
+}
+
+/// The representation of a color scale.
+/// The first `ColorStop` (position 0.0) defines the left end color.
+/// The last `ColorStop` (position 1.0) defines the right end color.
+#[derive(Debug, Clone)]
+pub struct ColorScale {
+    color_stops: Vec<ColorStop>,
+}
+
+impl ColorScale {
+    /// Create an empty `ColorScale`.
+    pub fn empty() -> Self {
+        Self {
+            color_stops: Vec::new(),
+        }
+    }
+
+    /// Add a `Color` at the given position.
+    pub fn add_stop(&mut self, color: Color, position: Fraction) -> &mut Self {
+        let color_stop = ColorStop { color, position };
+
+        let next_index = self
+            .color_stops
+            .iter()
+            .position(|c| position.value() <= c.position.value());
+
+        let index = match next_index {
+            Some(0) => 0,
+            Some(next) => next - 1,
+            None => self.color_stops.len(),
+        };
+
+        self.color_stops.insert(index, color_stop);
+
+        self
+    }
+
+    /// Get the color at the given position using the mixing function.
+    ///
+    /// Note:
+    /// - No color is returned if position isn't between two color stops or the `ColorScale` is empty.
+    pub fn sample(
+        &self,
+        position: Fraction,
+        mix: &Box<dyn Fn(&Color, &Color, Fraction) -> Color>,
+    ) -> Option<Color> {
+        let left_stop = self
+            .color_stops
+            .iter()
+            .rev()
+            .find(|c| position.value() >= c.position.value());
+
+        let right_stop = self
+            .color_stops
+            .iter()
+            .find(|c| position.value() <= c.position.value());
+
+        if left_stop.is_none() || right_stop.is_none() {
+            return None;
+        }
+
+        let left_stop = left_stop.expect("left stop");
+        let right_stop = right_stop.expect("right stop");
```

Comment:
Another option would be:
``` rust
match (left_stop, right_stop) {
  (Some(left_stop), Some(right_stop)) => {
    …
    Some(color)
  },
  _ => None
}
```

that would avoid the early `return` and the `.expect` calls. But I guess it's a matter of style. Your choice..